src/What4/ExprHelpers.hs

```diff
@@ -973,36 +1005,72 @@ simplifyBVOpInner sym go app = case app of
 -- shifts, pushing slices through arithmetic operations, and dropping
 -- zero-extensions where possible.
 simplifyBVOps ::
-  forall sym t solver fs tp.
+  forall sym m t solver fs tp.
+  IO.MonadUnliftIO m =>
   sym ~ (W4B.ExprBuilder t solver fs) =>
   sym ->
   W4.SymExpr sym tp ->
-  IO (W4.SymExpr sym tp)
-simplifyBVOps sym outer = do
+  m (W4.SymExpr sym tp)
+simplifyBVOps sym outer = simplifyBVOps' sym (\_ e2 -> return e2) outer
+
+
+simplifyBVOps' ::
+  forall sym m t solver fs tp.
+  IO.MonadUnliftIO m =>
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  (forall tp'. W4.SymExpr sym tp' ->  W4.SymExpr sym tp' -> m (W4.SymExpr sym tp')) {- ^ double-check simplification step -} ->
+  W4.SymExpr sym tp ->
+  m (W4.SymExpr sym tp)
+simplifyBVOps' sym simp_check outer = IO.withRunInIO $ \inIO -> do
   cache <- W4B.newIdxCache
   let
-    else_ :: forall tp'. W4.SymExpr sym tp' -> IO (W4.SymExpr sym tp')
-    else_ e = case e of
-      W4B.AppExpr a0 -> do
-        a0' <- W4B.traverseApp go (W4B.appExprApp a0)
-        if (W4B.appExprApp a0) == a0' then return e
-        else W4B.sbMakeExpr sym a0' >>= go
-      W4B.NonceAppExpr a0 -> do
-        a0' <- TFC.traverseFC go (W4B.nonceExprApp a0)
-        if (W4B.nonceExprApp a0) == a0' then return e
-        else W4B.sbNonceExpr sym a0' >>= go
-      _ -> return e      
-    
-    go :: forall tp'. W4.SymExpr sym tp' -> IO (W4.SymExpr sym tp')
-    go e = W4B.idxCacheEval cache e $ do
+    f :: forall tp'. W4B.App (W4B.Expr t) tp' -> m (Maybe (W4.SymExpr sym tp'))
+    f app = case simplifyBVOpInner sym (\e1 e2 -> inIO (simp_check e1 e2)) (\e' -> inIO (go e')) app of
+      Just g -> Just <$> liftIO g
+      Nothing -> return Nothing
+
+    go :: forall tp'. W4.SymExpr sym tp' -> m (W4.SymExpr sym tp')
+    go e = W4B.idxCacheEval cache e $ simplifyApp sym simp_check f e
+
+  inIO (go outer)
+
+simplifyApp ::
```

Comment:
What are the expected uses/semantics of the "double-check simplification step" - is this a full simplification validation?

Reply:
I wrapped up this check in a newtype and added some comments